pkg/ccl/crosscluster/logical/dead_letter_queue.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package logical
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
+	"github.com/cockroachdb/cockroach/pkg/repstream/streampb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+// TODO(azhu): Define WriteType as built-in database enum.
+type WriteType int
+
+const (
+	Insert WriteType = iota
+	Delete
+)
+
+func (t WriteType) String() string {
+	switch t {
+	case Insert:
+		return "Insert"
+	case Delete:
+		return "Delete"
+	default:
+		return fmt.Sprintf("Unrecognized WriteType(%d)", int(t))
+	}
+}
+
+type ConflictResolutionType int
+
+const (
+	Unresolved ConflictResolutionType = iota
+	ManuallyResolved
+	Applied
+	Ignored
+)
+
+type DeadLetterQueueClient interface {
+	Create() error
+
+	Log(
+		ctx context.Context,
+		ingestionJobID int64,
+		KVBatch []streampb.StreamEvent_KV,
+		rp RowProcessor,
+		conflictReason string,
+	) error
+}
+
+type TableDeadLetterQueueClient struct {
```

Comment:
nit: call this a `LoggingDeadLetterQueue`? We can create a separate struct when the time comes for a table dead letter queue.